### ISSUE TITLE
handling postgres deadlock timeouts.

### ISSUE THREAD

**Issue by chesty** [zammad/zammad#2114]
### Infos:

* Used Zammad version: 2.5.x
* Installation method (source, package, ..): zammad-docker-compose

### Expected behavior:

* zammad catches deadlock timeouts and re-queues job for a retry at a later time 

### Actual behavior:

* with imap zammad makes the mail unprocessable, with ldap sync zammad changes the account to inactive.

### Steps to reproduce the behavior:

I think there are 2 issues. 1 is handling deadlock timeouts, the other is why the deadlock timeout happened.

with handling deadlock timeouts, to simulate the issue, if updates to the organisation table normally take up to (eg) 50ms, set postgres deadlock timeouts smaller than that, let say 40ms, load up the server a bit so some updates start taking longer than 40ms and run an ldap sync and collect a bunch of emails and observe what happens when the deadlock timeout is triggered. 

I've set the postgres deadlock timeout to 15 seconds, which is really large for the average time queries take to run. Most queries take less than 50ms to run, when it's busy sql updates take a little longer occasionally, between 100-200ms.

The last week zammad has been perfect with no issues, last night I turned ldap sync back on, today I started seeing postgres dealock timeouts. It might be a coincidence?

Ideally the deadlock timeout should be the default 1 second and zammad retry after a timeout.

```
2018-07-05 14:37:12 AEST [37]: [5-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 LOG:  process 37 detected deadlock while waiting for ShareLock on transaction 1234486 afte
r 15000.083 ms
2018-07-05 14:37:12 AEST [37]: [6-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 DETAIL:  Process holding the lock: 2095. Wait queue: .
2018-07-05 14:37:12 AEST [37]: [7-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 CONTEXT:  while updating tuple (0,29) in relation "organizations"
2018-07-05 14:37:12 AEST [37]: [8-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 STATEMENT:  UPDATE "organizations" SET "updated_at" = '2018-07-05 04:36:57.636000' WHERE "
organizations"."id" = $1
2018-07-05 14:37:12 AEST [37]: [9-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 ERROR:  deadlock detected
2018-07-05 14:37:12 AEST [37]: [10-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 DETAIL:  Process 37 waits for ShareLock on transaction 1234486; blocked by process 2095.
        Process 2095 waits for ShareLock on transaction 1234487; blocked by process 37.
        Process 37: UPDATE "organizations" SET "updated_at" = '2018-07-05 04:36:57.636000' WHERE "organizations"."id" = $1
        Process 2095: UPDATE "users" SET "updated_at" = '2018-07-05 04:36:57.673000' WHERE "users"."id" = $1
2018-07-05 14:37:12 AEST [37]: [11-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 HINT:  See server log for query details.
2018-07-05 14:37:12 AEST [37]: [12-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 CONTEXT:  while updating tuple (0,29) in relation "organizations"
2018-07-05 14:37:12 AEST [37]: [13-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 STATEMENT:  UPDATE "organizations" SET "updated_at" = '2018-07-05 04:36:57.636000' WHERE "organizations"."id" = $1
2018-07-05 14:37:12 AEST [2095]: [5-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 LOG:  duration: 14962.678 ms  execute <unnamed>: UPDATE "users" SET "updated_at" = '2018-07-05 04:36:57.673000' WHERE "users"."id" = $1
2018-07-05 14:37:12 AEST [2095]: [6-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 DETAIL:  parameters: $1 = '116'
```

```
2018-07-05 14:48:30 AEST [2252]: [37-1] user=postgres,db=zammad_production,app=puma: cluster worker 0: 1 [zammad],client=172.18.0.7 LOG:  process 2252 detected deadlock while waiting for ShareLock on transaction 1242736 after 15000.114 ms
2018-07-05 14:48:30 AEST [2252]: [38-1] user=postgres,db=zammad_production,app=puma: cluster worker 0: 1 [zammad],client=172.18.0.7 DETAIL:  Process holding the lock: 38. Wait queue: .
2018-07-05 14:48:30 AEST [2252]: [39-1] user=postgres,db=zammad_production,app=puma: cluster worker 0: 1 [zammad],client=172.18.0.7 CONTEXT:  while updating tuple (0,53) in relation "organizations"
2018-07-05 14:48:30 AEST [2252]: [40-1] user=postgres,db=zammad_production,app=puma: cluster worker 0: 1 [zammad],client=172.18.0.7 STATEMENT:  UPDATE "organizations" SET "updated_at" = '2018-07-05 04:48:15.777000' WHERE "organizations"."id" = $1
2018-07-05 14:48:30 AEST [2252]: [41-1] user=postgres,db=zammad_production,app=puma: cluster worker 0: 1 [zammad],client=172.18.0.7 ERROR:  deadlock detected
2018-07-05 14:48:30 AEST [2252]: [42-1] user=postgres,db=zammad_production,app=puma: cluster worker 0: 1 [zammad],client=172.18.0.7 DETAIL:  Process 2252 waits for ShareLock on transaction 1242736; blocked by process 38.
        Process 38 waits for ShareLock on transaction 1242737; blocked by process 2252.
        Process 2252: UPDATE "organizations" SET "updated_at" = '2018-07-05 04:48:15.777000' WHERE "organizations"."id" = $1
        Process 38: UPDATE "users" SET "updated_at" = '2018-07-05 04:48:15.828000' WHERE "users"."id" = $1
2018-07-05 14:48:30 AEST [2252]: [43-1] user=postgres,db=zammad_production,app=puma: cluster worker 0: 1 [zammad],client=172.18.0.7 HINT:  See server log for query details.
2018-07-05 14:48:30 AEST [2252]: [44-1] user=postgres,db=zammad_production,app=puma: cluster worker 0: 1 [zammad],client=172.18.0.7 CONTEXT:  while updating tuple (0,53) in relation "organizations"
2018-07-05 14:48:30 AEST [2252]: [45-1] user=postgres,db=zammad_production,app=puma: cluster worker 0: 1 [zammad],client=172.18.0.7 STATEMENT:  UPDATE "organizations" SET "updated_at" = '2018-07-05 04:48:15.777000' WHERE "organizations"."id" = $1
2018-07-05 14:48:30 AEST [38]: [7-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 LOG:  duration: 14949.205 ms  execute <unnamed>: UPDATE "users" SET "updated_at" = '2018-07-05 04:48:15.828000' WHERE "users"."id" = $1
2018-07-05 14:48:30 AEST [38]: [8-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 DETAIL:  parameters: $1 = '116'
2018-07-05 14:48:30 AEST [11697]: [3-1] user=postgres,db=zammad_production,app=script/scheduler.rb,client=172.18.0.10 LOG:  disconnection: session time: 0:01:10.010 user=postgres database=zammad_production host=172.18.0.10 port=58084
```

Directly before and after the deadlock timeout, the slowest queries were taking between 50-100ms. So this appears to be caused by a programming issue, maybe with the order 2 threads are updating tables. I skimmed over an article weeks ago on why deadlocks happen and how to reduce the chances, and from memory the order the 2 threads update tables should be the same to reduce the chance of it occurring. But I'm no expert there and have full confidence in the zammad developers and community. Here's a different article I didn't read http://www.moioli.net/progetti/postgres-deadlocks-debugging-guidelines/  

It timed out after 15 seconds at 14:48:30, looking in the logs at 14:48:15, 15 seconds before the 15 second timeout, I see

```
I, [2018-07-05T14:48:13.882338 #1]  INFO -- : 2018-07-05T14:48:13+1000: [Worker(host:0cad51b628c7 pid:1)] Job BackgroundJobSearchIndex (id=6778527) COMPLETED after 1.7192
I, [2018-07-05T14:48:13.998434 #1]  INFO -- : 2018-07-05T14:48:13+1000: [Worker(host:0cad51b628c7 pid:1)] Job Transaction::BackgroundJob (id=6778528) RUNNING
I, [2018-07-05T14:48:15.608115 #1]  INFO -- : 2018-07-05T14:48:15+1000: [Worker(host:0cad51b628c7 pid:1)] Job Transaction::BackgroundJob (id=6778528) COMPLETED after 1.6096
I, [2018-07-05T14:48:15.610933 #1]  INFO -- : 2018-07-05T14:48:15+1000: [Worker(host:0cad51b628c7 pid:1)] Job Observer::Ticket::UserTicketCounter::BackgroundJob (id=6778529) RUNNING
I, [2018-07-05T14:48:16.960054 #1]  INFO -- : execute Import::OTRS.diff_worker (try_count 0)...
I, [2018-07-05T14:48:30.958922 #1]  INFO -- : Scheduler running...
I, [2018-07-05T14:48:30.965401 #1]  INFO -- : Running job thread for 'Check Channels' (Channel.fetch) status is: sleep
I, [2018-07-05T14:48:30.965486 #1]  INFO -- : Running job thread for 'Import OTRS diff load' (Import::OTRS.diff_worker) status is: sleep
I, [2018-07-05T14:48:30.965786 #1]  INFO -- : Started job thread for 'Process auto unassign tickets' (Ticket.process_auto_unassign)...
I, [2018-07-05T14:48:30.971293 #1]  INFO -- : execute Ticket.process_auto_unassign (try_count 0)...
I, [2018-07-05T14:48:30.975649 #1]  INFO -- :  ...stopped thread for 'Ticket.process_auto_unassign'
```

not sure if ldap sync was running then? don't think it was, it looks like it ran at 14:36 and took 1 minute to complete. Although it does generate a lot delayed_jobs, I'm not sure if they would have still been in the queue at 14:48.

I can email more logs if required. Happy to provide more info.

Yes I'm sure this is a bug and no feature request or a general question.

**Comment by thorsteneckel** [comment]
Wow! Thanks for the comprehensive writeup. It will probably take some time till we get here. Would be great if you could provide any information you have and will get in the future on this topic. We will get back to you as soon as we find the time to address this issue.

**Comment by chesty** [comment]
I haven't seen it happen since I reported it but I also turned LDAP sync off again. So I'm starting to think it's related to LDAP sync. I'll turn LDAP sync on over the weekend, see if I can trigger it. Are there any debug logs you would like me to turn on? If so I'll need instructions. The logs from the last time it happened are long gone.

This is my understanding of how deadlocks can happen, let's say there are 2 threads, thread A and thread B. A wants to update the organizations table, then the users table. B wants to update the users table, then the organizations table.
So A updates the organizations table at the same time B updates the users table, A has a sharelock on the organizations table and B has a sharelock on the users table. A now wants to update the users table, requests a lock, but B has it locked so A has to wait for B to release it, B wants to update the organizations table, but A has it locked, so B has to wait for A to release it. A and B are waiting for each other to release the lock they hold, and the deadlock timer times out.

Maybe ldap sync is updating tables in a different order to imap?

Happy to help with extra logs. I don't know ruby and I'm not familiar with the codebase otherwise I'd dig in and see if I could work out what order the organizations and users table are being updated in various jobs.
